### PR TITLE
[Traffic Manager] Adding new preview api-version for Traffic Manager. This adds ONE new feature to Traffic Manager: HeatMap.

### PR DESCRIPTION
**Pull request by hrkulkarMsft**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [X] The title of the PR is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [X] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [X] If applicable, the PR references the bug/issue that it fixes.
- [X] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [X] My spec meets the review criteria:
  - [X] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [X] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Comment by anuchandy** [comment]
@ravbhatnagar this is new new API version, please review and sign-off.

**Comment by anuchandy** [comment]
@hrkulkarMsft please update `readme.md` file to have a tag for this new api-version.

**Comment by hrkulkarMsft** [comment]
Should the readme.md still try to build non-preview api? Or should I add?


``` yaml
openapi-type: arm
tag: package-2017-09-preview
```

### Tag: package-2017-09-preview

These settings apply only when `--tag=package-2017-09-preview` is specified on the command line.

``` yaml $(tag) == 'package-2017-09-preview'
input-file:
- Microsoft.Network/2017-09-01-preview/trafficmanager.json
```

**Comment by anuchandy** [comment]
@hrkulkarMsft, the readme file seems correct, it seems CI is failing because from the swagger you are referring `HeatMap-GET.json` but in the file system the file name is `HeatMap-Get.json`, please use the same casing in both places.

**Comment by anuchandy** [comment]
If you already have autorest installed I would recommend to run it locally against your swagger to catch linter errors, and fix them.

Assuming you are in the root folder of local clone of rest-api-spec, you can run this:

```
autorest --validation --azure-validator --message-format=json --input-file=./specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json
```

**Comment by hrkulkarMsft** [comment]
Thanks, my validator wasn't throwing that warning because git hadn't renamed the case remotely to match my local branch.

For Operations API -- our parent RP (Microsoft.Network) implements this should we still reference it in our Swagger?

**Comment by anuchandy** [comment]
For `Operations API` no need to reference to the parent network `swagger`. @ravbhatnagar please note that linter shows RPCViolation saying `Operations API` is missing but the same operation is implemented in the parent RP https://github.com/Azure/azure-rest-api-specs/tree/current/specification/network/resource-manager.

**Comment by anuchandy** [comment]
To repro semantic and model validation locally

1. install oav
     `npm install -g oav`
2. run

```
oav validate-spec ./specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json
oav validate-example ./specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json 
```

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 1]
```diff
@@ -684,6 +684,187 @@
...
+              "$ref": "#/definitions/RealUserMetricsKeyModel" 
+            }
+          },
+          "default": {
+            "description": "Default response. It will be deserialized as per the Error definition.",
+            "schema": {
+              "$ref": "#/definitions/CloudError"
+            }
+          }
+        },
```
Comment (anuchandy) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: Is this a long running operation? if yes mark it so using ` "x-ms-long-running-operation": true`

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 2]
```diff
@@ -684,6 +684,187 @@
...
+          "204": {
+            "description": "The Real User Metrics key does not exist. It could have been deleted on a previous request."
+          },
+          "default": {
+            "description": "Default response. It will be deserialized as per the Error definition.",
+            "schema": {
+              "$ref": "#/definitions/CloudError"
+            }
+          }
+        },
```
Comment (anuchandy) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: Is this a long running operation? if yes mark it so using ` "x-ms-long-running-operation": true`

**Comment by hrkulkarMsft** [comment]
Thanks Anu, I was able to successfully run these validations this time.

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 3]
```diff
@@ -684,6 +684,187 @@
...
+        "operationId": "RealUserMetricsKey_CreateOrUpdate",
+        "description": "Create or update a Traffic Manager Real User Metrics key.",
+        "parameters": [
+          {
+            "name": "parameters",
+            "in": "body",
+            "required": false,
+            "schema": {
+              "type": "string",
+              "enum": [""]
```
Comment (anuchandy) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: curious what exactly this `enum` is? if not used please remove.

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json (hunk 1 above)
Comment (hrkulkarMsft) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: No, it should be quite short.

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json (hunk 2 above)
Comment (hrkulkarMsft) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: No, it should be short-running.

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json (hunk 3 above)
Comment (hrkulkarMsft) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: Removed this.

**Comment by anuchandy** [comment]
@hrkulkarMsft regarding `TrackedResourceListByImmediateParent` RPCViolation - you mentioned that `in the case of this particular resource -- it could be very large`, if i understand correctly the number of HeapMap child resources associated with an instance of traffic manager profile resource can be huge. If that is the case, this could be a paged collection right? which can be expressed using  `x-ms-pageable` extension, https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/creating-swagger.md#Paging-x-ms-pageable.

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json (hunk 1 above)
Comment (anuchandy) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: thanks for confirming

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json (hunk 2 above)
Comment (anuchandy) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: thanks for confirming

**Comment by hrkulkarMsft** [comment]
@anuchandy Sorry, I should have been more specific. The payload can return close to 4MB in our preview version -- this will be right under ARM limit to keep it from needing pagination. If we decide to have multiple heatMap resources at some point then (i.e.: a history of a Profile's heatMap), I would imagine that a List operation could get large.

**Comment by anuchandy** [comment]
@hrkulkarMsft thanks for clarifying, got it so there will be only one `HeapMap` child instance associated with a traffic manager profile instance, i.e. this is not a collection. I would let Gaurav comment on this. @ravbhatnagar we have `RPCViolation` warning in this case `TrackedResourceListByImmediateParent`, please share your thoughts.

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 4]
```diff
@@ -0,0 +1,1446 @@
...
+          }
+        },
+        "x-ms-examples": {
+          "HeatMap-GET": {
+            "$ref": "./examples/HeatMap-GET.json"
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Network/trafficManagerRealUserMetricsKey": {
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: What is this key? Is it a secret? Shouldn't this be obtained via a POST?

Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: The API signature is not correct for a GET or a PUT. it needs a segment for the resource name.

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 5]
```diff
@@ -0,0 +1,1446 @@
...
+          "x-ms-client-name": "operationResult",
+          "readOnly": true,
+          "description": "The result of the operation or request."
+        }
+      },
+      "description": "The result of the request or operation."
+    },
+    "EndpointProperties": {
+      "properties": {
+        "targetResourceId": {
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: How is this different from the "id" property in "Resource" model?

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 6]
```diff
@@ -0,0 +1,1446 @@
...
+            "name": "EndpointMonitorStatus",
+            "modelAsString": true
+          }
+        },
+        "minChildEndpoints": {
+          "type": "integer",
+          "format": "int64",
+          "description": "The minimum number of endpoints that must be available in the child profile in order for the parent profile to be considered available. Only applicable to endpoint of type 'NestedEndpoints'."
+        },
+        "geoMapping": {
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: Is this a fixed finite set? Consider exposing this list through an API or make an enum. Or is this list same as the supported azure locations?

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 7]
```diff
@@ -0,0 +1,1446 @@
...
+            "required": true,
+            "type": "string",
+            "description": "The name of the Traffic Manager endpoint to be updated."
+          },
+          {
+            "name": "parameters",
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/Endpoint"
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: Are all properties in the model patchable?

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 8]
```diff
@@ -0,0 +1,1446 @@
...
+          "NameAvailabilityTest_NameAvailablePOST21": {
+            "$ref": "./examples/NameAvailabilityTest_NameAvailable-POST-example-21.json"
+          },
+          "NameAvailabilityTest_NameNotAvailablePOST23": {
+            "$ref": "./examples/NameAvailabilityTest_NameNotAvailable-POST-example-23.json"
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/trafficmanagerprofiles": {
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: camelcase resource type name -> trafficManagerProfiles

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 9]
```diff
@@ -0,0 +1,1446 @@
...
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "The Traffic Manager Profile was deleted successfully.",
+            "schema": {
+              "$ref": "#/definitions/DeleteOperationResult"
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: there should not be  a response body for delete operation

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 10]
```diff
@@ -0,0 +1,1446 @@
...
+            "required": true,
+            "type": "string",
+            "description": "The name of the Traffic Manager profile."
+          },
+          {
+            "name": "parameters",
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/Profile"
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: All properties are patchable?

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 11]
```diff
@@ -0,0 +1,1446 @@
...
+          "description": "The DNS settings of the Traffic Manager profile."
+        },
+        "monitorConfig": {
+          "$ref": "#/definitions/MonitorConfig",
+          "description": "The endpoint monitoring settings of the Traffic Manager profile."
+        },
+        "endpoints": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/Endpoint"
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: Are you sure you want to return the entire Endpoint resource in the response of a Profile? You can just return the resource Ids for the endpoints.

Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: You will have to do a linked access check to make sure user has access to these resources (endpoints).

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 12]
```diff
@@ -0,0 +1,1446 @@
...
+        "path": {
+          "type": "string",
+          "description": "The path relative to the endpoint domain name used to probe for endpoint health."
+        },
+        "intervalInSeconds": {
+          "type": "integer",
+          "format": "int64",
+          "description": "The monitor interval for endpoints in this profile. This is the interval at which Traffic Manager will check the health of each endpoint in this profile."
+        },
+        "timeoutInSeconds": {
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: This should be ISO8601 format

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 13]
```diff
@@ -0,0 +1,1446 @@
...
+        "port": {
+          "type": "integer",
+          "format": "int64",
+          "description": "The TCP port used to probe for endpoint health."
+        },
+        "path": {
+          "type": "string",
+          "description": "The path relative to the endpoint domain name used to probe for endpoint health."
+        },
+        "intervalInSeconds": {
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: ISO 8601 format datetime?

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 14]
```diff
@@ -0,0 +1,1446 @@
...
+        },
+        "target": {
+          "type": "string",
+          "description": "The domain name or IP Address to which this endpoint refers to."
+        },
+        "resourceId": {
+          "type": "string",
+          "description": "The ARM Resource ID of this Traffic Manager endpoint."
+        },
+        "endpointId": {
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: Is this not the arm resource id of the endpoint

**Review by ravbhatnagar — CHANGES_REQUESTED** [review]
@hrkulkarMsft - please take a look at this feedback.

File: specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json [hunk 15]
```diff
@@ -1106,6 +1287,123 @@
...
+          "description": "The query experiences produced in this HeatMap calculation.",
+          "items": {
+            "$ref": "#/definitions/QueryExperience"
+          }
+        }
+      },
+      "description": "Class representing a Traffic Manager HeatMap properties."
+    },
+    "HeatMapEndpoint": {
+      "properties": {
```
Comment (ravbhatnagar) [inline specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json]: lets keep just resourceid

**Comment by hrkulkarMsft** [comment]
@ravbhatnagar : Submitted a new iteration in the review with the changes we discussed.


•	{heatMapsType}: can now only be “default” in value.
•	Removed endpoint-identifying properties.
•	Added query parameters.
•	Removed App Rum to get HeatMap through for now, and will open separate review.

I’ll take note of the general API feedback to fix, but for this iteration would it be possible to only fix the HeatMap implementation, and then update the existing API comments later?

**Comment by ravbhatnagar** [comment]
@hrkulkarMsft - sounds good.

**Comment by anuchandy** [comment]
RubyCoden is failing https://travis-ci.org/Azure/azure-rest-api-specs/jobs/268142263, checking with the codegen owners.

**Comment by hrkulkarMsft** [comment]
Hey, is there anything I need to do to unblock codegen?

**Comment by anuchandy** [comment]
hi @hrkulkarMsft sorry for the delay. No action required from your side now, we are tracking the code-gen issue here https://github.com/Azure/azure-sdk-for-ruby/issues/944. Merging this PR.

@ravbhatnagar as i mentioned above though there is a RPCViolation `OperationsAPIImplementation` reported by linter, the API is implemented in the parent Network swagger.

**Comment by hrkulkarMsft** [comment]
Cool, thanks Anu! Changed the PR to better reflect these changes since the other feature was moved to a separate PR.

**Pull request merged**